bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -260,151 +259,116 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
             case TypeTags.TYPE_REFERENCED_TYPE_TAG:
                 return getConvertibleTypes(inputValue, ((BTypeReferenceType) targetType).getReferredType(), varName,
-                        isFromJson, unresolvedValues, errors, allowAmbiguity);
+                        isFromJson, unresolvedValues, errors);
             case TypeTags.TYPEDESC_TAG:
                 return getConvertibleTypes(inputValue, ((BTypedescType) targetType).getConstraint(), varName,
-                        isFromJson, unresolvedValues, errors, allowAmbiguity);
+                        isFromJson, unresolvedValues, errors);
             default:
                 if (TypeChecker.checkIsLikeType(inputValue, targetType, true)) {
-                    convertibleTypes.add(targetType);
+                    return Set.of(targetType);
                 }
         }
-        return convertibleTypes;
+        return new LinkedHashSet<>();
     }
 
-    private static Set<Type> getDirectlyConvertibleTypesInUnion(Object inputValue, Type inputValueType,
-                                                                BUnionType targetType) {
-        boolean isInputValueSimpleBasicType = isSimpleBasicType(inputValueType);
-        Set<Type> directlyConvertibleTypes = new LinkedHashSet<>();
-        for (Type memType : targetType.getMemberTypes()) {
-            if (!isInputValueSimpleBasicType && TypeChecker.checkIsLikeType(inputValue, memType, false)) {
-                directlyConvertibleTypes.add(memType);
-                continue;
+    public static Set<Type> getFiniteTypeConvertibleTypes(Object inputValue, BFiniteType targetFiniteType,
+                                                          String varName, boolean isFromJson, List<String> errors,
+                                                          List<TypeValuePair> unresolvedValues) {
+        if (targetFiniteType.valueSpace.size() == 1) {
+            Type valueType = getType(targetFiniteType.valueSpace.iterator().next());
+            if (!isSimpleBasicType(valueType) && valueType.getTag() != TypeTags.NULL_TAG) {
+                return getConvertibleTypes(inputValue, valueType, varName, isFromJson, unresolvedValues,
+                        errors);
+            }
+        }
+        Type inputValueType = TypeChecker.getType(inputValue);
+        Set<Object> finiteTypeValueSpace = targetFiniteType.valueSpace;
+        for (Object valueSpaceItem : finiteTypeValueSpace) {
+            if (inputValue == valueSpaceItem) {
+                return Set.of(inputValueType);
+            }
+            if (TypeChecker.isFiniteTypeValue(inputValue, inputValueType, valueSpaceItem, false)) {
+                return Set.of(TypeChecker.getType(valueSpaceItem));
             }
-            if (inputValueType == memType || isIntegerSubtypeAndConvertible(inputValue, memType)) {
-                return Set.of(memType);
+        }
+        for (Object valueSpaceItem : finiteTypeValueSpace) {
```

Comment:
Can we reuse the above `for` loop by keeping the first convertible type (if there is any) inside a set and return the set outside the `for` loop?

Reply:
We can do, but I think `isFiniteTypeValue` and `getConvertibleTypes` methods are costly than the for loop. If the source value is of the target type, calls to those methods will be in vain. Hence I think it is better to have the expected sequential order.